crates/sui-core/src/validator_client_monitor/monitor.rs

```diff
@@ -0,0 +1,299 @@
+        let operation_str = match feedback.operation {
+            OperationType::Submit => "submit",
+            OperationType::Effects => "effects",
+            OperationType::HealthCheck => "health_check",
+        };
+
+        // Update metrics
+        if let Some(actual_latency) = feedback.latency {
+            self.metrics
+                .operation_latency
+                .with_label_values(&[&validator_str, operation_str])
+                .observe(actual_latency.as_secs_f64());
+        }
+
+        if feedback.success {
+            self.metrics
+                .operation_success
+                .with_label_values(&[&validator_str, operation_str])
+                .inc();
+        } else {
+            self.metrics
+                .operation_failure
+                .with_label_values(&[&validator_str, operation_str])
+                .inc();
+        }
+
+        // Update client stats
+        let mut client_stats = self.client_stats.write();
+        client_stats.record_interaction_result(feedback);
+    }
+
+    /// Handle epoch changes by cleaning up stale validator data.
+    ///
+    /// Called when the network transitions to a new epoch. Removes client-observed
+    /// data for validators that are no longer in the active committee to prevent
+    /// memory leaks and ensure scores are only calculated for current validators.
+    pub fn on_epoch_change(&self, epoch: EpochId) {
+        let authority_agg = self.authority_aggregator.load();
+        if authority_agg.committee.epoch == epoch {
+            return;
+        }
+
+        let current_validators: HashSet<_> =
+            authority_agg.authority_clients.keys().cloned().collect();
+
+        let mut client_stats = self.client_stats.write();
+        client_stats.refresh_validator_set(&current_validators);
+    }
+
+    /// Select validators based on client-observed performance with shuffled top k.
+    ///
+    /// First selects the top `k` validators by performance score, shuffles them,
+    /// then appends the remaining validators ordered by score. The selection considers:
+    /// - Current client-observed scores (reliability + latency)
+    /// - Exclusion status (temporarily excluded validators are filtered out)
+    /// - Committee membership (only selects from the provided committee)
+    ///
+    /// We need to pass the current committee here because it is possible
+    /// that the fullnode is in the middle of a committee change when this
+    /// is called, and we need to maintain an invariant that the selected
+    /// validators are always in the committee passed in.
+    ///
+    /// We shuffle the top k validators to avoid the same validator being selected
+    /// too many times in a row and getting overloaded.
+    ///
+    /// Returns a vector containing:
+    /// 1. The top `k` validators by score (shuffled)
+    /// 2. The remaining validators ordered by score (not shuffled)
+    pub fn select_shuffled_preferred_validators(
```

Comment:
Maybe in another change, I think this can take an OperationType too. The scoring can use the specified operation's stats plus health check stats.  One validator not receiving a submit request shouldn't affect other requests' selection.